File: src/main/java/shuaicj/hobby/http/proxy/netty/HttpProxyServer.java
package shuaicj.hobby.http.proxy.netty;

import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.handler.logging.LogLevel;
import io.netty.handler.logging.LoggingHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class HttpProxyServer {
    Logger logger = LoggerFactory.getLogger(this.getClass());

    private int port = 1080;

    public static void main(String[] args) {
        new HttpProxyServer().start();
    }

    public void start() {
        new Thread(() -> {
            logger.info("HttpProxyServer started on port: {}", port);
            EventLoopGroup bossGroup = new NioEventLoopGroup(1);
            EventLoopGroup workerGroup = new NioEventLoopGroup();
            try {
                ServerBootstrap b = new ServerBootstrap();
                b.group(bossGroup, workerGroup)
                        .channel(NioServerSocketChannel.class)
                        .handler(new LoggingHandler(LogLevel.DEBUG))
                        .childHandler(new HttpProxyChannelInitializer())
                        .bind(port).sync().channel().closeFuture().sync();
            } catch (InterruptedException e) {
                logger.error("shit happens", e);
            } finally {
                bossGroup.shutdownGracefully();
                workerGroup.shutdownGracefully();
            }
        }).start();
    }
}

File: src/main/scala/com/github/alikemalocalan/tunnel/HttpProxyServer.scala
package com.github.alikemalocalan.tunnel

import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.EventLoopGroup
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioServerSocketChannel
import io.netty.handler.logging.{LogLevel, LoggingHandler}
import org.slf4j.LoggerFactory

import scala.util.{Failure, Success, Try}

object HttpProxyServer extends App with Config {
  val logger = LoggerFactory.getLogger(this.getClass)

  start()

  def start(): Unit =
    new Thread { () =>

      logger.info("HttpProxyServer started on port: {}", port)
      val bossGroup = new NioEventLoopGroup(threadCount)
      val workerGroup: EventLoopGroup = new NioEventLoopGroup(1)

      Try {
        new ServerBootstrap()
          .group(bossGroup, workerGroup)
          .channel(classOf[NioServerSocketChannel])
          .handler(new LoggingHandler(LogLevel.INFO))
          .childHandler(new HttpProxyChannelInitializer())
          .bind(port)
          .sync()
          .channel()
          .closeFuture()
          .sync()
      } match {
        case Success(_) =>
          bossGroup.shutdownGracefully()
          workerGroup.shutdownGracefully()
        case Failure(exception) => logger.error("shit happens", exception);
      }

    }.start()

}
